src/pprof/ddprof_pprof.cc

```diff
@@ -34,54 +32,77 @@ std_string_2_slice_c_char(const std::string &str) {
 #define SLICE_LITERAL(str)                                                     \
   (struct ddprof_ffi_Slice_c_char) { .ptr = (str), .len = sizeof(str) - 1 }
 
+#define PPROF_MAX_LABELS 5
+
 static ddprof_ffi_Slice_c_char ffi_empty_char_slice(void) {
   return (struct ddprof_ffi_Slice_c_char){.ptr = (NULL), .len = 0};
 }
 
-static const struct ddprof_ffi_ValueType s_sample_count = {
-    .type_ = SLICE_LITERAL("sample"),
-    .unit = SLICE_LITERAL("count"),
-};
-
-static const struct ddprof_ffi_ValueType s_cpu_time = {
-    .type_ = SLICE_LITERAL("cpu-time"),
-    .unit = SLICE_LITERAL("nanoseconds"),
-};
-
-static const struct ddprof_ffi_ValueType s_cpu_sample = {
-    .type_ = SLICE_LITERAL("cpu-sample"),
-    .unit = SLICE_LITERAL("count"),
-};
-
-DDRes pprof_create_profile(DDProfPProf *pprof, const PerfOption *options,
-                           unsigned nbOptions) {
-  // Create one value
-  struct ddprof_ffi_ValueType perf_value_type[MAX_TYPE_WATCHER + 1];
-  unsigned value_index = 0;
-  perf_value_type[value_index++] = s_sample_count;
-
-  /* Add one value type per watcher */
-  for (; value_index < nbOptions + 1; ++value_index) {
-    ddprof_ffi_Slice_c_char perf_value_slice = {.ptr = options->label,
-                                                .len = strlen(options->label)};
-    ddprof_ffi_Slice_c_char perf_unit = {.ptr = options->unit,
-                                         .len = strlen(options->unit)};
-
-    perf_value_type[value_index].type_ = perf_value_slice;
-    perf_value_type[value_index].unit = perf_unit;
+DDRes pprof_create_profile(DDProfPProf *pprof, DDProfContext *ctx) {
+  // Figure out which sample_type_ids are used by active watchers
+  const PerfWatcher *watchers = ctx->watchers;
+  size_t num_watchers = ctx->num_watchers;
+  bool active_ids[DDPROF_PWT_LENGTH] = {};
+  struct ddprof_ffi_ValueType perf_value_type[DDPROF_PWT_LENGTH];
+  for (unsigned i = 0; i < num_watchers; ++i) {
+    int this_id = watchers[i].sample_type_id;
+    if (this_id < 0 || this_id == DDPROF_PWT_NOCOUNT || this_id >= DDPROF_PWT_LENGTH) {
```

Comment:
Could this happen?
If this is the case, then you probably need to check also on aggregation side that `sample_id` and `sample_type_pv[sample_id]` are both valid.

Reply:
I don't _think_ it could happen in the current code, but the boundary check is useful for preventing segfaults and DDPROF_PWT_NOCOUNT (which _should_ be the 0 value always, but there's no way to enforce that in the caller of the component).  Generally speaking, I feel like this is actually overly defensive, but at the same time this check is a boundary check on the pprof rather than the samples.

Not sure.  What do you think?  Overly defensive?  In the meantime, I agree that if we're being defensive, we might as well defend both of our castles and add a compatible check in the aggregation function--good idea.

Reply:
[Minor] I don't have a definite opinion about this (asserting pre-conditions vs defensively checking them). When modules are internal, I tend to prefer assertions.
I think what matters is to be consistent. For example here if you check for `sample_type_id` validity, then the check should also be done line 97. Or perhaps the function should just return an error, on the first check failure.